packages/zoe/src/contracts/priceAggregatorChainlink.js

```diff
@@ -0,0 +1,754 @@
+   * @param {bigint} lastReportedRound
+   * @param {bigint} lastStartedRound
+   * @param {bigint} latestSubmission
+   * @param {number} index
+   */
+  const makeOracleStatus = (
+    startingRound,
+    endingRound,
+    lastReportedRound,
+    lastStartedRound,
+    latestSubmission,
+    index,
+  ) => {
+    return {
+      startingRound,
+      endingRound,
+      lastReportedRound,
+      lastStartedRound,
+      latestSubmission,
+      index,
+    };
+  };
+
+  /**
+   *
+   * @param {PriceQuoteValue} quote
+   */
+  const authenticateQuote = async quote => {
+    const quoteAmount = AmountMath.make(quote, quoteKit.brand);
+    const quotePayment = await E(quoteKit.mint).mintPayment(quoteAmount);
+    return harden({ quoteAmount, quotePayment });
+  };
+
+  const { notifier } = makeNotifierKit();
+  const zoe = zcf.getZoeService();
+
+  /**
+   * @typedef {Object} OracleRecord
+   * @property {(timestamp: Timestamp) => Promise<void>=} querier
+   * @property {number} lastSample
+   */
+
+  /** @type {Store<Instance, Set<OracleRecord>>} */
+  const instanceToRecords = makeStore('oracleInstance');
+
+  /**
+   * @param {Object} param0
+   * @param {number} [param0.overrideValueOut]
+   * @param {Timestamp} [param0.timestamp]
+   */
+  const makeCreateQuote = ({ overrideValueOut, timestamp } = {}) =>
+    /**
+     * @param {PriceQuery} priceQuery
+     */
+    function createQuote(priceQuery) {
+      // Sniff the current baseValueOut.
+      const valueOutForUnitIn =
+        overrideValueOut === undefined
+          ? lastValueOutForUnitIn // Use the latest value.
+          : overrideValueOut; // Override the value.
+      if (valueOutForUnitIn === undefined) {
+        // We don't have a quote, so abort.
+        return undefined;
+      }
+
+      /**
+       * @param {Amount} amountIn the given amountIn
+       */
+      const calcAmountOut = amountIn => {
+        const valueIn = AmountMath.getValue(amountIn, brandIn);
+        return AmountMath.make(
+          floorDivide(multiply(valueIn, valueOutForUnitIn), unitIn),
+          brandOut,
+        );
+      };
+
+      /**
+       * @param {Amount} amountOut the wanted amountOut
+       */
+      const calcAmountIn = amountOut => {
+        const valueOut = AmountMath.getValue(amountOut, brandOut);
+        return AmountMath.make(
+          ceilDivide(multiply(valueOut, unitIn), valueOutForUnitIn),
+          brandIn,
+        );
+      };
+
+      // Calculate the quote.
+      const quote = priceQuery(calcAmountOut, calcAmountIn);
+      if (!quote) {
+        return undefined;
+      }
+
+      const {
+        amountIn,
+        amountOut,
+        timestamp: theirTimestamp = timestamp,
+      } = quote;
+      AmountMath.coerce(amountIn, brandIn);
+      AmountMath.coerce(amountOut, brandOut);
+      if (theirTimestamp !== undefined) {
+        return authenticateQuote([
+          { amountIn, amountOut, timer, timestamp: theirTimestamp },
```

Comment:
This is how we use the non-ERef form of the timer as an identity:

```suggestion
          { amountIn, amountOut, timer: timerPresence, timestamp: theirTimestamp },
```